librbd: support bps throttle and throttle read and write seperately.

yangdongsheng opened the pull request:
Signed-off-by: Dongsheng Yang <dongsheng.yang@easystack.cn>

yangdongsheng commented:
@junxfl  FYI, you can test this for what you want.

yangdongsheng commented:
We need to make sure the limit on read will not affect the writing in the same ImageRequestWQ.

junxfl commented:
@yangdongsheng  I add this patch to "https://github.com/yangdongsheng/ceph.git", branch is rbd_qos_read_write_bps.  it doesn't work as expected, I testd as below:
Jobs: 1 (f=1): [m(1)] [20.0% done] [95KB/135KB/0KB /s] [23/33/0 iops] [eta 00m:12s]
Jobs: 1 (f=1): [m(1)] [26.7% done] [79KB/43KB/0KB /s] [19/10/0 iops] [eta 00m:11s] 
Jobs: 1 (f=1): [m(1)] [33.3% done] [71KB/55KB/0KB /s] [17/13/0 iops] [eta 00m:10s]
Jobs: 1 (f=1): [m(1)] [40.0% done] [0KB/0KB/0KB /s] [0/0/0 iops] [eta 00m:09s]    
Jobs: 1 (f=1): [m(1)] [46.7% done] [163KB/63KB/0KB /s] [40/15/0 iops] [eta 00m:08s]
Jobs: 1 (f=1): [m(1)] [53.3% done] [43KB/47KB/0KB /s] [10/11/0 iops] [eta 00m:07s] 
Jobs: 1 (f=1): [m(1)] [60.0% done] [99KB/75KB/0KB /s] [24/18/0 iops] [eta 00m:06s]
^Cbs: 1 (f=1): [m(1)] [66.7% done] [99KB/95KB/0KB /s] [24/23/0 iops] [eta 00m:05s]

There are 3 metadata on this image:

Key                           Value 
conf_rbd_qos_iops_limit       120   
conf_rbd_qos_read_iops_limit  20    
conf_rbd_qos_write_iops_limit 50    
[root@node79 build]# cat fio-conf/config.iops.mix 
[global]   
ioengine=rbd  
clientname=admin  
pool=rbd  
rbdname=block 
bs=4k
runtime=15
iodepth=32 
rw=randrw
rwmixread=50
[test2]

yangdongsheng commented:
@junxfl don't use mix with fio for this test. because fio will send IO in 50:50. You can test this as you did with LIO+TCMU. One job for read and one job for write. 

I did a test with nbd as below:
```
[root@atest-guest fio]# ./fio examples/aio-read.fio --eta-newline 1
file1: (g=0): rw=randwrite, bs=(R) 4096B-4096B, (W) 4096B-4096B, (T) 4096B-4096B, ioengine=libaio, iodepth=4
file1: (g=0): rw=randread, bs=(R) 4096B-4096B, (W) 4096B-4096B, (T) 4096B-4096B, ioengine=libaio, iodepth=4
fio-2.99-dirty
Starting 2 processes
Jobs: 2 (f=2): [w(1),r(1)][0.2%][r=800KiB/s,w=400KiB/s][r=200,w=100 IOPS][eta 32m:43s]
Jobs: 2 (f=2): [w(1),r(1)][0.2%][r=800KiB/s,w=400KiB/s][r=200,w=100 IOPS][eta 36m:19s]
^Cbs: 2 (f=2): [w(1),r(1)][0.3%][r=800KiB/s,w=400KiB/s][r=200,w=100 IOPS][eta 38m:06s]
```

rbd-nbd map test

```
# cat examples/aio-read.fio 
; Read 4 files with aio at different depths
[global]
ioengine=libaio
buffered=0
bs=4K
filename=/dev/nbd0

[file1]
iodepth=4
rw=randwrite

[file1]
iodepth=4
rw=randread

# rbd image-meta list test
2018-04-25 08:11:24.450 7fcae3a35980 -1 WARNING: all dangerous and experimental features are enabled.
2018-04-25 08:11:24.489 7fcae3a35980 -1 WARNING: all dangerous and experimental features are enabled.
2018-04-25 08:11:24.520 7fcae3a35980 -1 WARNING: all dangerous and experimental features are enabled.
There are 3 metadata on this image:

Key                           Value 
conf_rbd_qos_iops_limit       0     
conf_rbd_qos_read_iops_limit  200   
conf_rbd_qos_write_iops_limit 100 
```

yangdongsheng commented:
@dillaman  Do you have time to review this PR?

dillaman commented on src/librbd/io/Types.h:
```diff
@@ -11,6 +11,18 @@
+#define RBD_QOS_WRITE_BPS_THROTTLE			1 << 5
+
+#define RBD_QOS_BPS_MASK	(RBD_QOS_BPS_THROTTLE | RBD_QOS_READ_BPS_THROTTLE | RBD_QOS_WRITE_BPS_THROTTLE)
+#define RBD_QOS_IOPS_MASK	(RBD_QOS_IOPS_THROTTLE | RBD_QOS_READ_IOPS_THROTTLE | RBD_QOS_WRITE_IOPS_THROTTLE)
```
Nit: used anywhere?

yangdongsheng replied:
Yup, such as
https://github.com/ceph/ceph/pull/21635/files#diff-82401c66599ad9652ae4be08000e5814R84

We need to check the flag with these mask to decide return ```spec->extents_length();``` (BPS tokens) or 1 (IOPS tokens).

dillaman commented on src/librbd/io/ImageRequestWQ.h:
```diff
@@ -95,8 +100,18 @@ class ImageRequestWQ
-  TokenBucketThrottle *iops_throttle;
+  std::atomic<unsigned> m_io_throttled { 0 };
+
+  std::list<uint64_t> m_throttle_flags = {
```
Nit: no need for this to be a member variable

dillaman commented on src/librbd/io/ImageRequestWQ.cc:
```diff
@@ -576,25 +581,89 @@ void ImageRequestWQ<I>::set_require_lock(Direction direction, bool enabled) {
+	  throttle->get<ImageRequestWQ<I>, ImageDispatchSpec<I>,
+	      &ImageRequestWQ<I>::handle_throttle_ready>(
+		tokens, this, item, flag)) {
+      return true;
```
Should the item by applied to all possible throttles in the first-pass so that it doesn't enter the back of the line when it's requeued? Perhaps the ```ImageDispatchSpec::was_throttled```/```ImageDispatchSpec::set_throttled``` logic should be reversed so that you default to all throttle bits set and you clear them if the throttle isn't required or when the throttle expires in ```handle_throttle_ready```

yangdongsheng replied:
Sorry I am afraid I did not catch all your point here :( .

But I did not set the all throttle bits for each request, instead, only check the throttles with m_qos_enabled_flag, because I consider this route is a hot route in IO path, I wish we can use as less time as possible here.

Could you help to explain more with your comment? Thanx a lot.

dillaman replied:
You need to avoid a race where a throttle completes while you are still iterating through throttles so that's why I was suggesting that it might be either to invert the default initialization state of the ```was/set_throttled``` bitmask (i.e. start w/ the assumption that throttling was required and turn it off bit-by-bit if it wasn't).

yangdongsheng replied:
Hmm, but the needs_throttle() and handle_throttle_ready() can't be called at the same time. I mean, we call needs_throttle() in _void_dequeue() when the request is in first-pass or requeued, the request is still in queue. But the handle_throttle_ready() is called when the request is already dequeued and throttled. So I think there is no race between them.

I think this is not what you want to say, I still did not get your point, sorry for that. could you please give me an example for it. Thanx a lot. :)

dillaman replied:
Yup -- you have two throttles active (rate and IOPS) and the request is throttled by both. The ```handle_throttle_ready``` will be called twice for the same request.

yangdongsheng replied:
Yes,  it would be called twice, but there is a flag to specify what we need to handle in handle_throttle_ready(), just something like that:
```
we have an image test with bps and iops throttle.


ImageRequestWQ:                                                                 BPS_Throttle                                    IOPS_Throttle
        |
        |---_void_dequeue()
                |
                |--needs_throttle()
                        |
                        |--throttle->get() <flag is for bps>         ---->   void TokenBucketThrottle::add_tokens()
                                                                                |
                                                                                |--b.ctx->complete(0);
        |
        |--handle_throttle_ready(flag) <flag is for bps>             <----
                |
                |---requeue(request)

        |
        |--_void_dequeue()
                |
                |--needs_throttle()
                        |
                        |--throttle->get() <flag is for iops>                                                   ---->   void TokenBucketThrottle::add_tokens()
                                                                                                                                |
                                                                                                                                |--b.ctx->complete(0);
        |
        |--handle_throttle_ready(flag) <flag is for iops>                                                       <----
                |
                |---requeue(request)

        |
        |--_void_dequeue()
                |
                |--needs_throttle() <no need to throttle now>
```

But, BTW, I found another problem here, if we get enough tokens, I forget to set_throttled(). so there is a fix as below:
```
diff --git a/src/librbd/io/ImageRequestWQ.cc b/src/librbd/io/ImageRequestWQ.cc
index dad7b2a..bbca232 100644
--- a/src/librbd/io/ImageRequestWQ.cc
+++ b/src/librbd/io/ImageRequestWQ.cc
@@ -654,11 +654,16 @@ bool ImageRequestWQ<I>::needs_throttle(ImageDispatchSpec<I> *item) {
 
     throttle = t.second;
     tokens = item->tokens_requested(flag);
-    if (!item->was_throttled(flag) &&
-         throttle->get<ImageRequestWQ<I>, ImageDispatchSpec<I>,
+
+    if (item->was_throttled(flag))
+      continue;
+
+    if (throttle->get<ImageRequestWQ<I>, ImageDispatchSpec<I>,
              &ImageRequestWQ<I>::handle_throttle_ready>(
                tokens, this, item, flag)) {
       return true;
+    } else {
+      item->set_throttled(flag);
     }
   }
   return false;
```

dillaman replied:
.... but getting back to the original comment, I think you should apply all applicable throttles in the initial pass so that a request that needs to be throttled by more than one throttle doesn't need to wait at the back of the second (or third ...) throttle queue after it's requeued.

dillaman commented on src/librbd/io/ImageRequestWQ.cc:
```diff
@@ -605,15 +674,12 @@ void *ImageRequestWQ<I>::_void_dequeue() {
-    // dequeue the throttled item and block future IO
+    // dequeue the throttled item
     ThreadPool::PointerWQ<ImageDispatchSpec<I> >::_void_dequeue();
-    ++m_io_blockers;
```
Need to come up w/ a plan where flushes cannot be re-ordered ahead of writes. Right now, if a write is throttled, a flush in the queue can proceed and flush "nothing" and complete before the write proceeds.

yangdongsheng replied:
Wow, good point! I missed the flush logic.  thanx a lot @dillaman

yangdongsheng replied:
Hi, @dillaman , I am thinking to make flush as an IO barrier to make sure flush is waiting for writes. But I have another question, should we guarantee the order of read and write to the same location?

yangdongsheng replied:
@dillaman After a second thought, maybe we don't need a special logic for flush. I checked qemu and kernel about the IO throttle. 
Qemu: https://github.com/qemu/qemu/blob/master/block/throttle.c#L104
```
static int coroutine_fn throttle_co_preadv(BlockDriverState *bs,
                                           uint64_t offset, uint64_t bytes,
                                           QEMUIOVector *qiov, int flags)
{

    ThrottleGroupMember *tgm = bs->opaque;
    throttle_group_co_io_limits_intercept(tgm, bytes, false);

    return bdrv_co_preadv(bs->file, offset, bytes, qiov, flags);
}

static int coroutine_fn throttle_co_pwritev(BlockDriverState *bs,
                                            uint64_t offset, uint64_t bytes,
                                            QEMUIOVector *qiov, int flags)
{
    ThrottleGroupMember *tgm = bs->opaque;
    throttle_group_co_io_limits_intercept(tgm, bytes, true);

    return bdrv_co_pwritev(bs->file, offset, bytes, qiov, flags);
}
... ...

static int throttle_co_flush(BlockDriverState *bs)
{
    return bdrv_co_flush(bs->file->bs);
}
```
We can see that read and write will be throttled by ```throttle_group_co_io_limits_intercept(tgm, bytes, true);```. But in flush, it did nothing check and just call the real flush ```bdrv_co_flush(bs->file->bs);```

kernel: https://github.com/torvalds/linux/blob/master/block/blk-core.c#L2249
```
static noinline_for_stack bool
generic_make_request_checks(struct bio *bio)
{
	struct request_queue *q;
	int nr_sectors = bio_sectors(bio);
	blk_status_t status = BLK_STS_IOERR;
	char b[BDEVNAME_SIZE];
... ...
	if (!blkcg_bio_issue_check(q, bio))
		return false;
...
```
then https://github.com/torvalds/linux/blob/master/include/linux/blk-cgroup.h#L683
```
static inline bool blkcg_bio_issue_check(struct request_queue *q,
					 struct bio *bio)
{
	struct blkcg *blkcg;
	struct blkcg_gq *blkg;
	bool throtl = false;

	rcu_read_lock();
	blkcg = bio_blkcg(bio);

... ...

	throtl = blk_throtl_bio(q, blkg, bio);

```
There are no logic to make flush wait the writes complete.

I mean, maybe the flush means make sure the all ```returned``` write operations flushed to disk rather than all ```queued``` write operations.  Example:
```
Thread A                                Thread B

write A and returned.
flush                                   Writing B

And we got the "write B" request first in workqueue.

In this case, the flush request should only guarantee that A
will be flushed to disk. About B, it maybe flushed if it completed
before flush start, or not if it didn't put his data into disk cache,
that's not guaranteed.

But yes, we need a locking in this case to avoid
this kind of race in this example, I just put it as
an example here.
``` 
I did not find an official Semantics for flush operation. Just learn this above from reading the code of qemu and kernel. That's your opinion? thanx

dillaman replied:
@yangdongsheng Sorry for the delay -- I've been side-tracked. 

> I did not find an official Semantics for flush operation. Just learn this above from reading the code of qemu and kernel. That's your opinion? thanx

I concur that this is the correct semantics so you are correct that you don't need to worry about throttling the flush. librbd internally just tracks the flush against all previous IOs as a way to handle IO guarding (e.g. blocking writes or completing all IOs when preparing to close the image).

yangdongsheng replied:
Yup, seems we need to override the drain() in ImageRequestWQ to complete all IOs when prepareing to close. :) thanx @dillaman

yangdongsheng commented:
Okey, I think I got your point now.

currently, I am handling the throttles one by one. but you think we can
apply them at one pass. So we should not return true if we get blocked by
one of the throttle in needs_throttle();

Hmmm, I got it. so there are different ways to do it. I need to think 
more about
which one is better. thanx for your explanation.

On 05/15/2018 09:49 AM, Jason Dillaman wrote:
> dillaman commented on this pull request.
>
>
>
>> +  uint64_t tokens = 0;
> +  uint64_t flag = 0;
> +  TokenBucketThrottle* throttle = nullptr;
> +
> +  for (auto t : m_throttles) {
> +    flag = t.first;
> +    if (!(m_qos_enabled_flag & flag))
> +      continue;
> +
> +    throttle = t.second;
> +    tokens = item->tokens_requested(flag);
> +    if (!item->was_throttled(flag) &&
> +	  throttle->get<ImageRequestWQ<I>, ImageDispatchSpec<I>,
> +	      &ImageRequestWQ<I>::handle_throttle_ready>(
> +		tokens, this, item, flag)) {
> +      return true;
>
> .... but getting back to the original comment, I think you should apply all applicable throttles in the initial pass so that a request that needs to be throttled by more than one throttle doesn't need to wait at the back of the second (or third ...) throttle queue after it's requeued.
>

yangdongsheng commented:
Hi @dillaman  I pushed some fixes on this branch, please help to take a look. If these looks good to you, I will squash them. thanx

dillaman commented on src/librbd/io/ImageDispatchSpec.h:
```diff
@@ -161,9 +169,9 @@ class ImageDispatchSpec {
   Request m_request;
   int m_op_flags;
   ZTracer::Trace m_parent_trace;
+  uint64_t m_throttled_flag = 0;
```
Nit: ```std::atomic<uint64_t>``` to avoid races w/ bitwise OR ops

dillaman commented on src/librbd/io/ImageDispatchSpec.h:
```diff
@@ -129,11 +129,18 @@ class ImageDispatchSpec {
+    m_throttled_flag |= flag;
+  }
+
+  bool was_all_throttled() {
```
Nit: ```was``` implies singular value, ```were``` for plural.

dillaman commented on src/librbd/io/ImageRequestWQ.h:
```diff
@@ -72,10 +72,20 @@ class ImageRequestWQ
   void set_require_lock(Direction direction, bool enabled);
 
   void apply_qos_iops_limit(uint64_t limit);
+  void apply_qos_bps_limit(uint64_t limit);
```
Nit: instead of 6 helper methods, perhaps just one that passes the limit and the ```RBD_QOS_XYZ``` flag.

dillaman commented:
@yangdongsheng Since ```TokenBucketThrottle``` isn't templated, probably the easiest thing to do would be to add a ```template<typename ImageCtxT> struct TypeTraits { typedef ::TokenBucketThrottle TokenBucketThrottle }``` to ```librbd/io/Types.h```. Then in ```test_mock_io_ImageRequestWQ.cc``` you could use template specialization to create a ```TypeTraits<MockTestImageCtx>``` that typedefs ```TokenBucketThrottle``` to some new mock class where you can control the invocation of the ```get``` call and ```handle_throttle_ready``` callback. That would allow you to create a test case where an IO is injected and throttled (by the test case) before the test case invokes the throttle callback.

runsisi commented on src/librbd/ImageCtx.cc:
```diff
@@ -771,7 +771,12 @@ class SafeTimerSingleton : public SafeTimer {
         "rbd_mirroring_replay_delay", false)(
         "rbd_skip_partial_discard", false)(
-	"rbd_qos_iops_limit", false);
+	"rbd_qos_iops_limit", false)(
```
i think unify tab & space is better :)

yangdongsheng replied:
@runsisi  thanx for your comment, yes, they are aligned actually, and this is just shown not aligned in github review mode, you can pull it and read it in vi, that's all aligned and unified.

yangdongsheng commented:
@dillaman  Hi jason, I pushed a commit for unittest case, please help to take a look. thanx

runsisi commented on src/librbd/io/ImageDispatchSpec.cc:
```diff
@@ -82,11 +121,27 @@ void ImageDispatchSpec<I>::fail(int r) {
 
+template <typename I>
+uint64_t ImageDispatchSpec<I>::extents_length() {
+    uint64_t length = 0;
```
indentation should be 2 spaces

dillaman commented on src/test/librbd/io/test_mock_ImageRequestWQ.cc:
```diff
@@ -306,5 +327,29 @@ TEST_F(TestMockIoImageRequestWQ, RefreshError) {
+  InSequence seq;
+  MockImageRequestWQ mock_image_request_wq(&mock_image_ctx, "io", 60, nullptr);
+
+  mock_image_request_wq.apply_qos_limit(1, RBD_QOS_IOPS_THROTTLE);
```
I haven't run it myself to confirm, but is this test showing that we don't initially populate the token bucket upon startup (i.e. this test will pause for 1 second)? If that's the case, perhaps we should always initially populate the bucket so that QoS user's don't always have a 1 second IO pause after opening the image.

yangdongsheng replied:
Hi @dillaman good point. I did a test for this:
```
[root@atest-guest build]# rbd bench --io-type read --io-size 4096 --io-threads 1 test1
bench  type read io_size 4096 io_threads 1 bytes 1073741824 pattern sequential
  SEC       OPS   OPS/SEC   BYTES/SEC
    1        20     21.02  86101.18
    2        40     20.50  83967.10
    3        60     20.33  83284.44
    4        80     20.25  82943.11
^C
[root@atest-guest build]# rbd image-meta list test1
There is 1 metadatum on this image:

Key                     Value 
conf_rbd_qos_iops_limit 20   
```
The result did not show the pause of 1 second after opening the image. But I agree that we need to initially populate the token bucket. There are two cases we will init a token. 

(1) constructer of Bucket:
```
    Bucket(CephContext *cct, const std::string& n, uint64_t m)
      : cct(cct), name(n),
        remain(m), max(m)
    {
    }
```
That's okey, we initialized the remain with m which means the capacity. 
(2) set_max()
```
void TokenBucketThrottle::Bucket::set_max(uint64_t m) {
  if (remain > m)
    remain = m;
  max = m;
}
```
There is a possibility in this case that when we change the state from non-limit to limit of 100, we only change the max from 0 to 100, but the remain is still 0 and wait for timer to put tokens. So I pushed a commit to fix this case.

dillaman commented on src/common/Throttle.cc:
```diff
@@ -685,7 +685,7 @@ uint64_t TokenBucketThrottle::Bucket::put(uint64_t c) {
 
 void TokenBucketThrottle::Bucket::set_max(uint64_t m) {
-  if (remain > m)
+  if (remain > m || m == 0)
```
Nit: should this be ```max == 0```?

dillaman commented:
@yangdongsheng see the make check test failure

yangdongsheng commented:
Thanx @dillaman  . That's a good news, means the unittest case works well and it can find some problem in logic.  This failure is because I changed the set_max() to solve the 1 second problem you mentioned. So the unittest case will not be blocked then.

I have changed the unittest case to make Bucket capacity to 1 and token_requested to 2. then we will be blocked anyway and it can test the full path. 

In addition, I added another unittest case for QosNoLimit, to check the other path without qos throttling.
